### ISSUE TITLE
[Feature]: Dynamic token in the same authentication API

### ISSUE THREAD

**Issue by felix-appsmith** [appsmithorg/appsmith#20100]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Summary

This function that allows our user to dynamically change the authentication token of the data source without the need to enter the data source to change it automatically, this function could look something like this.
![image](https://user-images.githubusercontent.com/114161539/214707378-fa5fc19b-e644-4d69-8f30-7d012e28411e.png)

### Why should this be worked on?

This function will allow the same data source to be more dynamic by being able to enter the same source with different tokens or if the data source has tokens that expire from time to time, the user would not have to change it manually.

**Comment by cbresson33** [comment]
I need the same capacity. I am using Magento API, which provides a token for a limited period of time. I need to call a first API to get the updated token and put it dynamically into the API data source.
This is mandatory to use appSmith with the Magento API.
I hope this option will be available soon.

**Comment by Nikhil-Nandagopal** [comment]
@cbresson33 can you point to the magento docs you are using to obtain the access token?

**Comment by Nikhil-Nandagopal** [comment]
@felix-appsmith the request here seems to simply need it to work with OAuth so this solution is incorrect and this should work today as well right?

**Comment by cbresson33** [comment]
Yes, using OAuth could be en option, but my business partner in Magento development does not agree to set it on (too complicated for him).

Using a token by directly writing its value in datasource works well. But it stops working when the token becomes invalid (after 20mn).

It would be easier for me to be able to change the token value dynamically at each request.

Thanks for your help.

Ch.

De : Nikhil Nandagopal ***@***.***>
Envoyé : lundi 20 mars 2023 13:31
À : appsmithorg/appsmith ***@***.***>
Cc : Christophe Bresson ***@***.***>; Mention ***@***.***>
Objet : Re: [appsmithorg/appsmith] [Feature]: Dynamic token in the same authentication API (Issue #20100)


@felix-appsmith<https://github.com/felix-appsmith> the request here seems to simply need it to work with OAuth so this solution is incorrect and this should work today as well right?

**Comment by cbresson33** [comment]
I use this API point with a POST call to obtain the token :  https://<MagentoAPI>/rest/all/V1/integration/admin/token<https://%3cMagentoAPI%3e/rest/all/V1/integration/admin/token>

ID and password are provided in the body of the request.

Here is the Magento doc about it : https://devdocs.magento.com/guides/v2.3/get-started/authentication/gs-authentication-token.html

Thanks for your help.

Regards,

Ch.

De : Nikhil Nandagopal ***@***.***>
Envoyé : lundi 20 mars 2023 09:46
À : appsmithorg/appsmith ***@***.***>
Cc : Christophe Bresson ***@***.***>; Mention ***@***.***>
Objet : Re: [appsmithorg/appsmith] [Feature]: Dynamic token in the same authentication API (Issue #20100)


@cbresson33<https://github.com/cbresson33> can you point to the magento docs you are using to obtain the access token?

**Comment by Nikhil-Nandagopal** [comment]
@cbresson33 in the authenticated API header, you can already configure your access token like in the image below to achieve this if you are manually storing and sending the access token. So if the API fails, you simply. need to fetch the new access token and store it using `{{storeValue('token', fetch_token.data.token)}}` and the authenticated API will send it in the headers

<img width="1397" alt="Screenshot 2023-03-20 at 7 45 34 PM" src="https://user-images.githubusercontent.com/3897254/226367274-8f90ae30-aba8-4d65-9881-d0c91bc32d4c.png">

**Comment by felix-appsmith** [comment]
Hello @Nikhil-Nandagopal , the function requested by the user is designed to facilitate the use case that was described in the previous comments. In addition to enabling this use case, the function can also support other use cases. For example, it can allow a single data source to be shared among different users who connect to it using different tokens. This can grant varying levels of privileges to different users on the same data source, which can simplify some processes and make others more efficient.

By using this function, users can connect to the same data source using a unique token that determines the level of access they have. For instance, some users may have read-only access, while others may have write access, allowing them to modify or update the data source. This can help in scenarios where multiple users need to access the same data source, but with different levels of privileges.

This function can be particularly useful for organizations that need to manage a single data source across multiple teams or departments. It can also be helpful for scenarios where different user roles require different levels of access to the same data source. Overall, this function can simplify data management processes, reduce the time required for data access, and make data more secure by controlling who can access it and to what extent.

**Comment by geekyme-fsmk** [comment]
Upvoting on this. My use case here is that my bearer tokens are obtained dynamically so I cannot hardcode a value at the data source level.
